### PR TITLE
Add volunteer voucher settings, batch linking, and bulk send

### LINKED ISSUE #107: Generate and send ticket vouchers to individual team members.
For many events, organisers would give volunteers a free or discounted ticket. We need a simple way to connect TeamShifts with our existing ticket voucher setup and allow organizers to send vouchers to members from TeamShifts message center. 

### 1. Organiser sets up a voucher batch in Tickets/Products/Voucher

Image

#### Add a Ticket Vouchers section below People on the left menu bar. This will open up:

**Example only**
 
Here you can configure free or discounted tickets for team members: 

☐ Enable volunteer vouchers
Voucher batch [ Select voucher batch ▾ ]
Vouchers remaining: 48 / 50

[ Create a voucher batch in Tickets → ]

**Notes:**
- The dropdown lists all existing voucher batches for this event
- A "Create a voucher batch in Tickets →" link takes the organiser directly to the right place in Tickets if they have not created one yet
- The remaining count shows how many unused codes are left in the selected batch

### 3. One voucher per volunteer
- When the organiser sends a voucher to a member via the Members page, TeamShifts picks one unused code from the selected batch and assigns it to that member. Each member gets exactly one unique code, no duplicates.

**This is just a mockup generated by AI as reference**

Image

## Key behaviour
- No batch selected If the organiser has not selected a voucher batch, the Send voucher button on the Members page is disabled with a message: "Configure a voucher batch in settings first.
- Batch runs out If all codes have been used, the Send voucher button is disabled with a warning: "Voucher batch is empty. Add more codes in Tickets → Vouchers.
- All volunteer vouchers appear in Tickets → Vouchers alongside all other vouchers. The organiser has a complete picture of voucher usage in one place.
- The organiser can change the selected batch at any time. Vouchers already sent to volunteers are not affected. They keep their existing codes.

## Expected outcome 
- Ticket Voucher section appears on the side bar below People section 
- Organiser can select any existing voucher batch from the event
- Helper link takes organiser to Tickets → Vouchers to create a batch if needed
- Remaining voucher count shows correctly from the selected batch
- Each volunteer receives exactly one unique code — no duplicates
- Batch empty warning appears when all codes are used

### LINKED ISSUE #108: Send ticket voucher code to team members
This is a follow-up of this #107 covering the manual Send voucher bulk action on the Members page. 

### How it works
- Organiser selects one or more members using the row checkboxes
- Bulk action bar appears at the top of the table
- Organiser clicks Send voucher
- A standalone email is sent to each selected member with their voucher code
- Voucher column status updates accordingly

### Voucher Email Template 
Subject: Your ticket voucher — {{event_name}}

Hi {{member_name}},

Here is your ticket voucher for {{event_name}}.

- Voucher code: {{voucher_code}}
- Claim your ticket here: {{ticket_claim_url}}

This voucher is valid for one ticket and can only be used once.

If you haven't yet claimed your shifts, please visit the shift schedule to sign up: {{shift_schedule_url}}

See you at the event!

Your {event_name} team

### Key behaviours

- Not sent → generates and sends a new voucher If the member has never received a voucher, a new unique voucher code is generated at the moment the button is clicked and sent in the email. Voucher status changes from "Not sent" to "Sent — not claimed".
- Sent — not claimed → resends the same voucher If the member already received a voucher but has not claimed it, the same existing voucher code is resent — not a new one. This prevents a volunteer having multiple active codes. Voucher status remains "Sent — not claimed".
- Claimed → skipped If the member has already claimed their voucher, they are skipped silently. No email is sent.
- After the bulk action A summary message appears at the top of the page: "Vouchers sent to 3 members. 1 skipped (already claimed)."
- Vouchers disabled If the organiser has not enabled vouchers in Ticket Vouchers section, the Send voucher button is hidden from the bulk action bar entirely.

### PR DESCRIPTION
**Pull request by MukundC25**
Connects TeamShifts with the existing ticket voucher system so organisers can send free/discounted ticket codes to accepted team members.

Closes #107 
Closes #108 

What's included:
- Ticket Vouchers settings page (sidebar entry below People) where the organiser enables vouchers and selects an existing voucher batch by tag
- Remaining/total voucher count display on the settings page
- Helper link to create a voucher batch in Tickets → Vouchers
- Voucher status column on the Members page (Not sent / Sent — not claimed / Claimed)
- Bulk "Send voucher" action with checkboxes and select-all
- One unique code per member — resends the same code if already sent, skips if already claimed
- Voucher email template with {voucher_code} and {ticket_claim_url} placeholders
- Batch-empty and no-batch-configured guard rails

Models: `VolunteerVoucherSettings`, `MemberVoucher`
Migration: `0020_volunteer_vouchers`

https://github.com/user-attachments/assets/01b97ebb-5197-4781-8abb-e50815ca9083

Follow-up items (not in this PR):
- Voucher email template is editable from the Email Templates page (the default template is auto-created on first use, but the Templates UI does not yet surface the VOUCHER_SENT role for editing)
- CSV export of member voucher assignments
- Individual "Send voucher" button per member row (current scope is bulk only)

## Summary by Sourcery

Enable organisers to configure and bulk-send unique ticket vouchers to accepted team members while tracking voucher lifecycle status.

New Features:
- Add organiser settings for enabling volunteer vouchers and selecting tagged ticket voucher batches, including batch availability counts and a link to create batches.
- Add bulk voucher sending from the Members page with member selection, voucher status tracking, and guard rails for missing or exhausted batches.
- Send voucher emails containing voucher codes and ticket claim links using a dedicated email template role and placeholders.

Enhancements:
- Persist one-to-one member voucher assignments so unclaimed vouchers can be resent and claimed vouchers are skipped.

Tests:
- Add service tests covering unique voucher allocation, resends, claimed-voucher skipping, empty batches, and email failures.

## Summary by CodeRabbit

* **New Features**
  * Added Ticket Vouchers settings to enable vouchers and select a voucher batch.
  * Added bulk voucher sending and resend support for eligible team members.
  * Added voucher status tracking: Not sent, Sent, and Claimed.
  * Added member selection controls and voucher status visibility.
  * Added configurable voucher-sent email notifications with voucher details and claim links.
  * Added voucher availability counts and a link to create voucher batches.
  * Added validation requiring a voucher batch when vouchers are enabled.

### CODE REVIEW

File: teamshifts/services/vouchers.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    result: SendResult = {"sent": 0, "resent": 0, "skipped_claimed": 0, "skipped_no_vouchers": 0}
+
+    try:
+        cfm = event.call_for_team_members
+    except CallForTeamMembers.DoesNotExist:
+        cfm = None
+
+    if cfm is None:
+        logger.warning("[TeamShifts] No CFM for event %s, cannot resolve voucher email template", event.slug)
+        return result
```
[the remark this answers is not in the document]
Comment (MukundC25) [reply teamshifts/services/vouchers.py:54]: Partially addressed. The empty-result case no longer produces a blank message, the view now falls back to "No vouchers were sent."

File: teamshifts/forms.py [hunk 2]
```diff
@@ -755,3 +755,37 @@ def __init__(self, *args, user=None, **kwargs):
...
+        label=_("Voucher batch"),
+        help_text=_("Select the voucher batch (tag) created in Tickets → Vouchers."),
+        widget=forms.Select(attrs={"class": "form-control"}),
+    )
+
+    def __init__(self, *args, event=None, **kwargs):
+        initial = kwargs.get("initial", {})
+        self.event = event
+        super().__init__(*args, **kwargs)
+
```
[the remark this answers is not in the document]
Comment (MukundC25) [reply teamshifts/forms.py:776]: fixed. Added `clean()` that errors if enabled is set without a `voucher_tag`

File: teamshifts/services/vouchers.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    context["ticket_claim_url"] = ticket_claim_url
+
+    subject = LazyI18nString(template.subject)
+    body = LazyI18nString(template.body)
+
+    try:
+        mail(
+            email=user.email,
+            subject=subject,
+            template=body,
```
[the remark this answers is not in the document]
Comment (MukundC25) [reply teamshifts/services/vouchers.py:141]: fixed.

File: teamshifts/views.py [hunk 4]
```diff
@@ -1659,6 +1673,21 @@ def get_context_data(self, **kwargs):
             "can_teamshifts_manage_applicants",
             request=self.request,
         )
+
+        voucher_settings = self._get_voucher_settings()
+        ctx["vouchers_enabled"] = bool(voucher_settings and voucher_settings.enabled and voucher_settings.voucher_tag)
+        ctx["vouchers_not_configured"] = not ctx["vouchers_enabled"]
+        ctx["voucher_batch_empty"] = (
+            voucher_settings.batch_remaining_count() == 0
```
[the remark this answers is not in the document]
Comment (MukundC25) [reply teamshifts/views.py]: fixed. `vouchers_not_configured` is now True only when vouchers are enabled but no batch is selected.

File: teamshifts/services/vouchers.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    resent: int
+    skipped_claimed: int
+    skipped_no_vouchers: int
+
+
+def allocate_and_send_vouchers(
+    event: Event,
+    settings: VolunteerVoucherSettings,
+    applications: Sequence[TeamMemberApplication],
+) -> SendResult:
```
[the remark this answers is not in the document]
Comment (MukundC25) [reply teamshifts/services/vouchers.py:37]: not addressed in this change

File: teamshifts/services/vouchers.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
```
Comment (Saksham-Sirohi) [inline teamshifts/services/vouchers.py:37]: Please add tests for this service: allocate unique codes, resend the same code when unclaimed, skip claimed vouchers, and the empty-batch guard. This is new ticket-code allocation with `select_for_update` concurrency, and there is no coverage for it yet.

**Review by Saksham-Sirohi — CHANGES_REQUESTED** [review]
Allocation/resend/mail-failure handling looks solid. Please add coverage for the new voucher service before merge.

File: teamshifts/services/vouchers.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    template = cfm.get_mail_template(EmailTemplateRoles.VOUCHER_SENT)
+    locale = event.settings.locale
+
+    for application in applications:
+        user = application.user
+        if not user or not user.email:
+            continue
+
+        with scope(event=event):
+            existing = MemberVoucher.objects.filter(application=application).select_related("voucher").first()
```
Comment (Aqil-Ahmad) [inline teamshifts/services/vouchers.py:65]: MemberVoucher.objects.filter(application=application) runs once per member inside the loop, as does refresh_claimed_status's refresh_from_db (models.py:781). For a bulk send that's 2-3 queries per member on top of the SMTP call. Prefetching all MemberVoucher rows for the selected applications up front would collapse it to one query. (Commit 678aca6 moved toward prefetched objects — looks like this path was missed.)

File: teamshifts/services/vouchers.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+            member_voucher.status = VoucherStatus.SENT
+            member_voucher.sent_at = now()
+            member_voucher.save(update_fields=["status", "sent_at"])
+            result["sent"] += 1
+
+    return result
+
+
+def _claim_next_voucher(settings: VolunteerVoucherSettings) -> Voucher | None:
+    """Pick and lock one unused voucher from the batch.
```
Comment (Aqil-Ahmad) [inline teamshifts/services/vouchers.py:104]: unbounded pk__in subquery. assigned_ids is evaluated as a subquery each call, and _claim_next_voucher is called once per member, so the exclusion set is recomputed per allocation.

File: teamshifts/static/teamshifts/members.js [hunk 9]
```diff
@@ -38,7 +38,39 @@ function restoreButtonChildren(button, originalChildren) {
     button.replaceChildren(...originalChildren);
 }
 
+function updateBulkActionBar() {
```
Comment (Aqil-Ahmad) [inline teamshifts/static/teamshifts/members.js:41]: selection is page-local. document.querySelectorAll(".member-checkbox:checked") only sees the current page. Worth checking whether that list paginates; if it does, the selection-store approach applies.

File: teamshifts/services/vouchers.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+        cfm = None
+
+    if cfm is None:
+        logger.warning("[TeamShifts] No CFM for event %s, cannot resolve voucher email template", event.slug)
+        return result
+
+    template = cfm.get_mail_template(EmailTemplateRoles.VOUCHER_SENT)
+    locale = event.settings.locale
+
+    for application in applications:
```
Comment (Aqil-Ahmad) [inline teamshifts/services/vouchers.py:59]: silent skip. Members with no user or no email are continued without landing in any SendResult bucket, so the summary message undercounts and the admin never learns those members were skipped. A skipped_no_email counter would close the loop.

**Review by Aqil-Ahmad — CHANGES_REQUESTED** [review]

File: teamshifts/services/vouchers.py (hunk 8 above)
Comment (MukundC25) [reply teamshifts/services/vouchers.py:125]: Fixed as a side-effect of comment 1

File: teamshifts/static/teamshifts/members.js (hunk 9 above)
Comment (MukundC25) [reply teamshifts/static/teamshifts/members.js]: Proper cross-page selection (server-side "select all matching query" or client-side store) is a meaningful feature on its own. Deferring to a follow-up.

File: teamshifts/services/vouchers.py (hunk 10 above)
Comment (MukundC25) [reply teamshifts/services/vouchers.py:76]: Fixed

**Review by Aqil-Ahmad — COMMENTED** [review]
Please see if any of these can be fixed here:
 
1. Its a bit confusing at first that how to send the vouchers, as the button only appears when you check the checkboxes. Keep it visible like bulk accept and reject buttons, and to the right with blue color.
Image

2. Sending vouchers takes some loading time which will become more noticable with many members. Seems intentional decision.
3. Only one voucher can be sent per member
4. the email doesn't appear in sent, I assume you will follow-up on this.

**Review by Aqil-Ahmad — APPROVED** [review]
